File: optimizeSx.py
#! /usr/bin/env python
"""
Name:
  optimizeSx  
Purpose:
  explore the presumed arbitrary cut off point for S_{1/2} by optimizing
    PTE(S_x) for random CMB realizations
Note:  
  Copi et al 2015 state: 
    "the statistical significance of the absence of large-angle correlations 
     is not particularly dependent either on the precise value of either limit" 
     (of the |C(theta)|^2 integral)
Uses:
  healpy
  legprodint (legendre product integral)
Inputs:

Outputs:

Modification History:
  Written by Z Knight, 2016.09.12

  Still need to optimize smica S_x just like others.
  
"""

import numpy as np
import matplotlib.pyplot as plt
import healpy as hp
import time # for measuring duration
from scipy.interpolate import interp1d

import get_crosspower as gcp # for loadCls
from numpy.polynomial.legendre import legval # for C_l -> C(theta) conversion
from numpy.polynomial.legendre import legfit # for C(theta) -> C_l conversion
from ispice import ispice
import legprodint
import sim_stats as sims # for getSMICA and getCovar


class Jmn():
  """
  Purpose:
    hold Jmn(x) for various x values and interpolate between them

  Note:
    I may drop this and do the interpolation with the S_x values instead


  Procedure:
    
  Inputs:

  Returns:
    
  """

  def __init__(self,lmax=100,new=True):
    if new:
      print 'new Jmn init running'
    else: 
      print 'old Jmn init running'
    
    nVals = 11#101
    self.Jmnx = np.empty([nVals,lmax+1,lmax+1])
    self.xvalues = np.linspace(0.25,0.75,nVals)
    for index, val in enumerate(self.xvalues):
      self.Jmnx[index] = legprodint.getJmn(endX=val,lmax=lmax,doSave=False)

  def getJmn(self,x):
    """
    Purpose:
      gets (interpolates?) Jmn(x)
    """
    print 'getJmn does not yet interpolate'

    return self.Jmnx



################################################################################
# testing code

def test(useCLASS=1,useLensing=0,classCamb=1,nSims=1000,lmax=100,lmin=2,
         newSMICA=False,newDeg=False):
  """
    code for testing the other functions in this module
    Inputs:
      useCLASS: set to 1 to use CLASS, 0 to use CAMB
        CLASS Cl has early/late split at z=50
        CAMB Cl has ISWin/out split: ISWin: 0.4<z<0.75, ISWout: the rest
        Note: CAMB results include primary in ISWin and ISWout (not as intended)
        default: 1
      useLensing: set to 1 to use lensed Cl, 0 for non-lensed
        default: 0
      classCamb: if 1: use the CAMB format of CLASS output, if 0: use CLASS format
        Note: parameter not used if useCLASS = 0
        default: 1
      nSims: the number of simulations to do for ensemble
        default: 1000
      lmax: the highest l to include in Legendre transforms
        default: 100
      lmin: the lowest l to include in S_{1/2} = CIC calculations
        default: 2
      newSMICA: set to True to recalculate SMICA results
        default: False
      newDeg: set to True to recalculate map and mask degredations
        (only if newSMICA is also True)
        default: False
  """
#  # first I want something that can create an array of Jmn(x) 
#  # this will be used to find S_x as well as help define histogram bins to store results
#  myJmn = Jmn(new=True)
#  Jmnx=myJmn.getJmn(0.5)
#  print 'Jmnx result: ',Jmnx.shape,Jmnx



  ##############################################################################
  # load theoretical power spectra

  # load data
  ell,fullCl,primCl,lateCl,crossCl = gcp.loadCls(useCLASS=useCLASS,useLensing=useLensing,
                                                  classCamb=classCamb)

  # fill beginning with zeros
  startEll = ell[0]
  ell      = np.append(np.arange(startEll),ell)
  fullCl   = np.append(np.zeros(startEll),fullCl)
  primCl   = np.append(np.zeros(startEll),primCl)
  lateCl   = np.append(np.zeros(startEll),lateCl)
  crossCl  = np.append(np.zeros(startEll),crossCl)

  # suppress C_2 to see what happens in enesmble
  suppressC2 = False
  suppFactor = 0.23 # from Tegmark et. al. 2003, figure 13 (WMAP)
  if suppressC2:
    fullCl[2] *= suppFactor
    primCl[2] *= suppFactor
    lateCl[2] *= suppFactor
    crossCl[2] *= suppFactor

  conv = ell*(ell+1)/(2*np.pi)
  #print ell,conv #ell[0]=2.0

  # apply beam and pixel window functions to power spectra
  #   note: to ignore the non-constant pixel shape, W(l) must be > B(l)
  #     however, this is not true for NSIDE=128 and gauss_beam(5')
  #   Here I ignore this anyway and proceed
  myNSIDE = 128 # must be same NSIDE as in sims.getSMICA function
  Wpix = hp.pixwin(myNSIDE)
  Bsmica = hp.gauss_beam(5./60*np.pi/180) # 5 arcmin
  WlMax = Wpix.size
  if WlMax < lmax:
    print 'die screaming!!!'
    return 0
  fullCl  =  fullCl[:WlMax]*(Wpix*Bsmica)**2
  primCl  =  primCl[:WlMax]*(Wpix*Bsmica)**2
  lateCl  =  lateCl[:WlMax]*(Wpix*Bsmica)**2
  crossCl = crossCl[:WlMax]*(Wpix*Bsmica)**2
  # note: i tried sims without this scaling, and results seemed the same at a glance


  ##############################################################################
  # load SMICA data, converted to C(theta), via SpICE

  if newSMICA:
    theta_i = 0.0 #degrees
    theta_f = 180.0 #degrees
    nSteps = 1800
    thetaArray2sp, C_SMICAsp, C_SMICAmaskedsp, S_SMICAnomasksp, S_SMICAmaskedsp = \
      sims.getSMICA(theta_i=theta_i,theta_f=theta_f,nSteps=nSteps,lmax=lmax,lmin=lmin,
               newSMICA=newSMICA,newDeg=newDeg,useSPICE=True)

  # filenames for SpICE to use
  # super lame that spice needs to read/write from disk, but here goes...
  ClTempFile  = 'tempCl.fits'
  mapTempFile = 'tempMap.fits'
  mapDegFile  = 'smicaMapDeg.fits' # this should have been created by sims.getSMICA
  maskDegFile = 'maskMapDeg.fits'  # this should have been created by sims.getSMICA
  
  ispice(mapDegFile,ClTempFile,maskfile1=maskDegFile,subav="YES",subdipole="YES")
  ClsmicaCut = hp.read_cl(ClTempFile)

  # find S_{1/2} for SMICA.  Should actually optimize but see what happens here first.
  myJmn = legprodint.getJmn(endX=0.5,lmax=lmax,doSave=False)
  Ssmica = np.dot(ClsmicaCut[lmin:lmax+1],np.dot(myJmn[lmin:,lmin:],
                  ClsmicaCut[lmin:lmax+1]))*1e24 #K^4 to microK^4


  ##############################################################################
  # create ensemble of realizations and gather statistics

  spiceMax = myNSIDE*3 # should be lmax+1 for SpICE
  ClEnsembleCut = np.zeros([nSims,spiceMax])
  simEll = np.arange(spiceMax)

  doTime = True # to time the run and print output
  startTime = time.time()
  for nSim in range(nSims):
    print 'starting masked Cl sim ',nSim+1, ' of ',nSims
    alm_prim,alm_late = hp.synalm((primCl,lateCl,crossCl),lmax=lmax,new=True)
    mapSim = hp.alm2map(alm_prim+alm_late,myNSIDE,lmax=lmax)
    hp.write_map(mapTempFile,mapSim)

    ispice(mapTempFile,ClTempFile,maskfile1=maskDegFile,subav="YES",subdipole="YES")
    ClEnsembleCut[nSim] = hp.read_cl(ClTempFile)

    doPlot = False#True
    if doPlot:
      gcp.showCl(simEll[:lmax+1],ClEnsembleCut[nSim,:lmax+1],
                  title='power spectrum of simulation '+str(nSim+1))

  if doTime: print 'time elapsed: ',int((time.time()-startTime)/60.),' minutes'


  ##############################################################################
  # create S(x) for each C_l, using interpolation
  
  nXvals = 181
  thetaVals = np.linspace(0,180,nXvals) # one degree intervals
  xVals = np.cos(thetaVals*np.pi/180)
  Jmnx = np.empty([nXvals,lmax+1,lmax+1])
  for index, xVal in enumerate(xVals):
    Jmnx[index] = legprodint.getJmn(endX=xVal,lmax=lmax,doSave=False)
  SxToInterpolate = np.empty(nXvals)

  # create list of functions
  dummy = lambda x: x**2
  SofXList = [dummy for i in range(nSims)]

  for nSim in range(nSims):
    print 'starting S(x) sim ',nSim+1,' of ',nSims
    for index,xVal in enumerate(xVals):
      SxToInterpolate[index] = np.dot(ClEnsembleCut[nSim,lmin:lmax+1],
          np.dot(Jmnx[index,lmin:,lmin:],ClEnsembleCut[nSim,lmin:lmax+1]))
    SofX = interp1d(xVals,SxToInterpolate)
    #SofXList = SofXList.append(SofX)
    # Apparently appending a function to an empty list is not allowed. Instead:
    SofXList[nSim] = SofX

    #print SofXList#[nSim]
    doPlot=False#True
    if doPlot:
      nplotx = (nXvals-1)*10+1
      plotTheta = np.linspace(0,180,nplotx)
      plotx = np.cos(plotTheta*np.pi/180)
      plotS = SofXList[nSim](plotx)
      plt.plot(plotx,plotS)
      plt.title('S(x) for simulation '+str(nSim+1))
      plt.show()

  doPlot = True
  if doPlot:
    for nSim in range(nSims):
      nplotx = (nXvals-1)*10+1
      plotTheta = np.linspace(0,180,nplotx)
      plotx = np.cos(plotTheta*np.pi/180)
      plotS = SofXList[nSim](plotx)
      plt.plot(plotx,plotS,label='sim '+str(nSim+1))
    #plt.legend()
    plt.title('S(x) for simulations')
    plt.show()



  ##############################################################################
  # create Pval(x) for each S(x), using ensemble
  # Pval: probability of result equal to or more extreme

  # create list of functions
  PvalOfXList = [dummy for i in range(nSims)]
  
  for nSim in range(nSims):
    print 'starting Pval(x) sim ',nSim+1,' of ',nSims

    def PvalOfX(x):
      nUnder = 0 # will also include nEqual
      nOver  = 0
      threshold = SofXList[nSim](x)
      for nSim2 in range(nSims):
         Sx = SofXList[nSim2](x)
         if Sx > threshold: 
           nOver  += 1
         else: 
           nUnder += 1
      return nUnder/float(nUnder+nOver)
    PvalOfXList[nSim] = PvalOfX

  
  ##############################################################################
  # find global minimum for each Pval(x)
  # simply use same xVals as above, at one degree intervals
  # if there are equal p-values along the range, the one with the highest xVal
  #   will be reported

  PvalMinima = np.empty(nSims)
  xValMinima = np.empty(nSims)
  for nSim in range(nSims):
    PvalOfX = PvalOfXList[nSim]
    #print 'function: ',PvalOfX
    PvalMinima[nSim] = PvalOfX(1.0)
    xValMinima[nSim] = 1.0

    Pvals = np.empty(nXvals)
    for index,xVal in enumerate(xVals): # will start from 1 and go down to -1
      myPval = PvalOfX(xVal)
      Pvals[index] = myPval
      if myPval < PvalMinima[nSim]:
        PvalMinima[nSim] = myPval
        xValMinima[nSim] = xVal
        #print 'nSim: ',nSim+1,', new x for minimum Pval: ',xVal

    doPlot = True
    if doPlot and np.random.uniform() < 0.1: #randomly choose about 1/10 of them
      plt.plot(xVals,Pvals)
      plt.vlines(xValMinima[nSim],0,1)
      plt.xlabel('x = cos(theta), min at '+str(xValMinima[nSim]))
      plt.ylabel('P-value')
      plt.title('P-values for simulation '+str(nSim+1)+' of '+str(nSims)+
          ', p_min = '+str(PvalMinima[nSim]))
      plt.xlim(-1.05,1.05)
      plt.ylim(-0.05,1.05)
      plt.show()


  # A MYSTERY!  Something about the following code causes Pvals to always take 
  #   the values of PvalOfXList[nSims](xVals)  WTF?  Omit for now. 
  #   Testing seems to indicate that PvalOfXList functions still have different
  #   locations in memory, but they all seem to be evaluating the same.
  #   However, when the previous block of code is copied to come again after
  #   this one, it behaves properly again.
  # see how well it did
  doPlot = False#True
  if doPlot:
    nPlots = 10
    for nPlot in range(nPlots):
      print 'plot ',nPlot+1,' of ',nPlots
      toPlot = nPlot#np.random.randint(0,high=nSims)
      #for nSim in range(nSims):
      Pvals = np.empty(nXvals)
      PvalOfX = PvalOfXList[nPlot]
      print 'function: ',PvalOfX
      for index, xVal in enumerate(xVals):
        Pvals[index] = PvalOfX(xVal)
        #print index,Pvals[index]
      #print Pvals
      plt.plot(xVals,Pvals)
      plt.vlines(xValMinima[toPlot],0,1)
      plt.xlabel('x = cos(theta), min at '+str(xValMinima[toPlot]))
      plt.ylabel('P-value')
      plt.title('P-values for simulation '+str(toPlot+1)+' of '+str(nSims))
      plt.show()


  ##############################################################################
  # create distribution of S(xValMinima)

  SxEnsembleMin = np.empty(nSims)
  for nSim in range(nSims):
    SxEnsembleMin[nSim] = SofXList[nSim](xValMinima[nSim])


  ##############################################################################
  # plot/print results

  
  print 'plotting S_x distribution... '
  myBins = np.logspace(2,7,100)
  plt.axvline(x=Ssmica,color='g',linewidth=3,label='SMICA masked')
  plt.hist(SxEnsembleMin, bins=myBins,histtype='step',label='cut sky')

  plt.gca().set_xscale("log")
  plt.legend()
  plt.xlabel('S_{1/2} (microK^4)')
  plt.ylabel('Counts')
  plt.title('S_{1/2} of '+str(nSims)+' simulated CMBs')
  plt.show()




  print 'step 3: profit'

if __name__=='__main__':
  test()

File: optimizeSx.c
#include <stdio.h>
/*
 * create S_x functions from input file
 * create Pval(x) for each S_x using ensemble
 * find global minimum for each Pval(x)
 *   -> can be used to create distribution S(xValMinima)
 *
 * compile with:
 *  gcc -O3 -fPIC -shared -o optimizeSx.so optimizeSx.c
 *
 * Modification History:
 *  Written by Z Knight, 2016.09.26
 *  Corrected 2d array indexing problem in optSx variable SxVecs:
 *    apparently passing python numpy arrays into SxVecs using 
 *    python's ctypes ndpointer inadvertantly flattened the array.
 *    Fixed by rewriting to expect 1d array; ZK, 2016.09.28
 *    Modified xstart value to avoid oddness near x=-1.0; ZK, 2016.09.30
 *    Added xStart and xEnd to optSx function to constrain search range; ZK, 2016.10.06
 *
 */

double linearInterp(double x1, double y1, double x2, double y2, double xMid) {
  /*
   * x1,y1: coordinates of left point
   * x2,y2: coordinates of right (x2>x1) point
   * xMid: point to interpolate at (x1<xMid<x2)
   *  (ok if outside range, but that would actually be extrapolation)
   */
  double slope;
  double intercept;
  slope = (y2-y1)/(x2-x1);
  intercept = y2-slope*x2;
  //printf("x1: %f, x2: %f, x2-x1: %f, y2: %f, y1: %f, y2-y1: %f\n",x1,x2,x2-x1,y2,y1,y2-y1);
  //printf("slope: %f, intercept: %f\n",slope,intercept);
  return slope*xMid+intercept;
}


double arrayLinearInterp(const double *xVec, const double *yVec, const double xInterp, size_t xSize) {
  /*
   * Note: not for use with negative y values, due to error return code -1
   * Inputs:
   *  xVec: an array of points to interpolate between, in decreasing order
   *  yVec: the corresponding y values
   *  xInterp: the point to interpolate at
   *    must be within range of highest and lowest values in xVec
   *  xSize: the number of elements in x array
   * Returns:
   *  the interpolated y value, or -1 if xInterp was outside range
   */
  // due to decreasing order of xVec, signs are not as expected:
  //if (xInterp < xVec[0] || xInterp > xVec[xSize-1]) {
  if (xInterp > xVec[0] || xInterp < xVec[xSize-1]) {
    printf("%f, %f, %f",xVec[0],xInterp,xVec[xSize-1]);
    puts("xInterp outside of interpolation range.  Exiting.");
    return -1;
  }
  size_t i = 0; //loop index
  while (i < xSize) {
    // due to decreasing order of xVec, signs are not as expected:
    if (xInterp >= xVec[i]) 
    //  return linearInterp(xVec[i],yVec[i],xVec[i+1],yVec[i+1],xInterp);
    //if (xInterp <= xVec[i+1]) 
      return linearInterp(xVec[i+1],yVec[i+1],xVec[i],yVec[i],xInterp);
    i++;
  }
  return -1; // this line is just to keep the compiler warnings quiet

}

double PvalOfX(double xVal, size_t nSim, const double *xVec, const double *SxVecs, size_t xSize, size_t nSims) {
  /*
   * Purpose: calculate p-values for each S_x, using ensemble
   *  Pval: probability of result equal to or more extreme
   * Inputs:
   *  xVal: the x value to calculate p-values for
   *  nSim: which of the simulations to evaluate for
   *    (this replaces the python version's list of functions)
   *  xVec, SxVecs, xSize: for forwarding to arrayLinearInterp
   *  nSims: number of sims in SxVecs
   * Returns:
   *  the p-value
   */
  int nUnder = 0;
  int nOver  = 0;
  // sending pointer to flattened array: nSim*xSize points to the [0] element in sim nSim 
  double threshold = arrayLinearInterp(xVec,&SxVecs[nSim*xSize],xVal,xSize);
  double mySx;  // for calculating Sx values
  for (size_t mySim = 0; mySim < nSims; mySim++) {
    mySx = arrayLinearInterp(xVec,&SxVecs[mySim*xSize],xVal,xSize);
    if (mySx > threshold) {
      nOver++;
      //printf("Over! mySx: %f, threshold: %f\n",mySx,threshold);
    }
    else                  {
      nUnder++;
      //printf("Under! mySx: %f, threshold: %f\n",mySx,threshold);
    }
  }
  //printf("nUnder: %d, nOver: %d\n",nUnder,nOver);
  return nUnder/(double)(nUnder+nOver);

}

void optSx(const double *xVec, size_t xSize, const double *SxVecs, size_t nSims, double xStart, double xEnd, int nSearch, double *PvalMinima, double *XvalMinima)  {
  /*
   * Name:
   *  optSx
   * Purpose:
   *  find global minimum for each Pval(x)
   * Procedure:
   *  simply use xVals defined on equal intervals and search for min. from left
   *  Note: if there are equal p-values along the range, the one with the
   *    lowest xVal will be reported
   * Inputs:
   *  xVec: an array of x points for S_x values
   *  xSize: the number of points in xVec array
   *  SxVecs: array of arrays of S_x values
   *    This can be thought of as a 2d array, but really it's flattened into 1d.
   *    Its length is nSims*xSize; its index is nSim*xSize+nX
   *  nSims: the number of simulations (rows) in SxVecs
   *  xStart: the start of the x range to search
   *    must be >= -1 and < xEnd.  If -1 is entered, -0.999 will be used 
   *      instead to avoid chaotic behavior of Pval(x) near x=-1
   *  xEnd: the end of the x range to search
   *    must be <= 1 and > xStart
   *  nSearch: the number of points to search for along x
   *  PvalMinima,XvalMinima: arrays of length nSims to contain P(x) and x values
   * Returns:
   *  void
   *  However, variables PvalMinma and XvalMinima will contain return values
   *
   */

  //for (size_t i = 0; i < xSize; i++) {
  //  printf("%.10e ",SxVecs[5*xSize+i]);
  //}
  //printf("\nThat was c's SxVecs[5] Press enter.\n");
  //getchar();

  // create empty pVal array and linspace x array
  //double myPvals[nSearch];
  double myPval;
  double myXvals[nSearch];
  //double xStart = -0.999; //-1.0; //start for range (-1 <= x <= 1)
  if (xStart == -1) {
    xStart = -0.999; //avoid chaotic Pval(x) near -1
  }
  double deltaX = xEnd - xStart;
  for (int n = 0; n < nSearch; n++) {
    myXvals[n] = xStart + deltaX*(n/(double)(nSearch-1));
  }

  //size_t nSim; // loop index
  for (size_t nSim = 0; nSim < nSims; nSim++) {
    printf("starting minimum Pval(x) search for sim %zd of %zd\n",nSim+1,nSims);
    PvalMinima[nSim] = PvalOfX(xStart,nSim,xVec,SxVecs,xSize,nSims);
    XvalMinima[nSim] = xStart;
    for (int n = 1; n < nSearch; n++) { // can omit 0 due to above PvalOfX

      myPval = PvalOfX(myXvals[n],nSim,xVec,SxVecs,xSize,nSims);
      //printf("nSim: %zd, n: %d, myPval: %f, PvalMinima[nSim]: %f\n",nSim,n,myPval,PvalMinima[nSim]);
      if (myPval < PvalMinima[nSim] && myXvals[n] > -0.999) {
        // -0.999 to avoid instabililility at endpoint
        PvalMinima[nSim] = myPval;
        XvalMinima[nSim] = myXvals[n];
      } 
    }    
    //printf("Finished sim %zd of %zd.  Press enter to continue\n",nSim+1,nSims);
    //getchar();
  }
  // return PvalMinima, XvalMinima; //nope, return in parameters instead

}


int main( int argc, char *argv[]) {
  /*
   * Originally this was going to be a program that took a file name as
   *  an argument, opened it, and wrote another file back to disk.
   * Instead, this program is now intended to be used as a shared object
   *  that is linked to from the optimizeSx.py python program.
   *
   * So, this main function does almost nothing, since it doesn't get used.
   *
   */

  if( argc == 2 ) {
    printf("The argument supplied is %s\n",argv[1]);
  }
  else if( argc > 2 ) {
    printf("Too many arguments supplied.\n");
  }
  else {
    printf("One argument expected.\n");
  }
}


